Signal/src/view controllers/MessagesViewController.m

```diff
@@ -495,50 +492,166 @@ - (void)setNavigationTitle
-
 {
+    UIImage *backImage = [UIImage imageNamed:@"NavBarBack"];
+    OWSAssert(backImage);
+    UIBarButtonItem *backItem = [[UIBarButtonItem alloc] initWithImage:backImage style:UIBarButtonItemStylePlain target:self action:@selector(backButtonPressed:)];
+    
+    const CGFloat kTitleVSpacing = 0.f;
+    if (!self.navigationBarTitleView) {
+        self.navigationBarTitleView = [UIView new];
+        [self.navigationBarTitleView addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self
+                                                                                                  action:@selector(navigationTitleTapped:)]];
+        
+        self.navigationBarTitleLabel = [UILabel new];
+        self.navigationBarTitleLabel.textColor = [UIColor whiteColor];
+        self.navigationBarTitleLabel.font = [UIFont ows_boldFontWithSize:18.f];
+        self.navigationBarTitleLabel.lineBreakMode = NSLineBreakByTruncatingTail;
+        [self.navigationBarTitleView addSubview:self.navigationBarTitleLabel];
+        
+        self.navigationBarSubtitleLabel = [UILabel new];
+        self.navigationBarSubtitleLabel.textColor = [UIColor colorWithWhite:0.85f alpha:1.f];
+        self.navigationBarSubtitleLabel.font = [UIFont ows_boldFontWithSize:9.f];
+        self.navigationBarSubtitleLabel.text = NSLocalizedString(@"MESSAGES_VIEW_TITLE_SUBTITLE",
+                                                                 @"The subtitle for the messages view title indicates that the title can be tapped to access settings for this conversation.");
+        [self.navigationBarTitleView addSubview:self.navigationBarSubtitleLabel];
+    }
+    
+    // We need to manually resize and position the title views;
+    // iOS AutoLayout doesn't work inside navigation bar items.
+    [self.navigationBarTitleLabel sizeToFit];
+    [self.navigationBarSubtitleLabel sizeToFit];
+    const CGFloat kShortScreenDimension = MIN([UIScreen mainScreen].bounds.size.width,
+                                              [UIScreen mainScreen].bounds.size.height);
+    // We want to leave space for the "back" button, the "timer" button, the "call" or
+    // "group settings" button, and all of the whitespace around these views.  There
+    // isn't a convenient way to calculate these in a navigation bar, so we just leave
+    // a constant amount of space which will be safe unless Apple makes radical changes
+    // to the appearance of the navigation bar.
+    int rightBarButtonItemCount = 0;
+    if ([self canCall]) {
+        rightBarButtonItemCount++;
+    }
+    if (disappearingMessagesConfiguration.isEnabled) {
+        rightBarButtonItemCount++;
+    }
+    CGFloat rightBarButtonSize = 0;
+    switch (rightBarButtonItemCount) {
+        case 0:
+            rightBarButtonSize = 65;
+            break;
+        case 1:
+            rightBarButtonSize = 100;
+            break;
+        default:
+            OWSAssert(0);
+        case 2:
```

Comment:
I put default here so that in production it would fall through to the largest defined case.